prepare.go

```diff
@@ -0,0 +1,451 @@
+package main
+
+import (
+	"bytes"
+	"compress/gzip"
+	"crypto/sha256"
+	"fmt"
+	"io"
+	"math"
+	"os"
+	"strings"
+	"time"
+)
+
+// Internal struct for hashing a simple single part file without compression
+type singlePartFileInfo struct {
+	Sha256 []byte
+	Size   int64
+}
+
+// Internal struct for hashing a simple single part file with compression
+type compressedSinglePartFileInfo struct {
```

Comment:
I was thinking I could probably merge all these internal structs, but I didn't want to need to know the nil value of each member type.